peps/pep-0750.rst

```diff
@@ -53,662 +34,602 @@ deferred pending further experience with f-strings. Work on this PEP was
+between the two types of elements in the sequence. 
 
-The ``format_spec`` is by default ``None`` if it is not specified in the tag string's
-corresponding interpolation.
+The detail exists because it allows for performance optimizations in template
+processing code. For example, a template processor could cache the static parts
+of the template and only reprocess the dynamic parts when the template is
+evaluated with different values. Access to the static parts can be done with
+``template.args[::2]``.
 
-Because the tag function is completely responsible for processing ``Decoded``
-and ``Interpolation`` values, there is no required interpretation for the format
-spec and conversion in an interpolation. For example, this is a valid usage:
 
-.. code-block:: python
-
-    html'<div id={id:int}>{content:HTML|str}</div>'
+Examples
+========
 
-In this case the ``format_spec`` for the second interpolation is the string
-``'HTML|str'``; it is up to the ``html`` tag to do something with the
-"format spec" here, if anything.
+All examples in this section of the PEP have fully tested reference implementations
+available in the public `pep750-examples <https://github.com/davepeck/pep750-examples>`_ 
+git repository.
 
-f-string-style ``=`` Evaluation
--------------------------------
 
-``mytag'{expr=}'`` is parsed to being the same as ``mytag'expr={expr}``', as
-implemented in the issue `Add = to f-strings for
-easier debugging <https://github.com/python/cpython/issues/80998>`_.
+Example: Implementing f-strings With t-strings
+----------------------------------------------
 
-Tag Function Arguments
-----------------------
+It is easy to "implement" f-strings using t-strings. That is, we can 
+write a function ``f(template: Template) -> str`` that processes a ``Template``
+in much the same way as an f-string literal, returning the same result:
 
-The tag function has the following signature:
 
 .. code-block:: python
 
-    def mytag(*args: Decoded | Interpolation) -> Any:
-        ...
+    name = "World"
+    value = 42.0
+    templated = t"Hello {name!r}, value: {value:.2f}"
+    formatted = f"Hello {name!r}, value: {value:.2f}"
+    assert f(templated) == formatted
 
-This corresponds to the following protocol:
+The ``f()`` function supports both conversion specifiers like ``!r`` and format
+specifiers like ``:.2f``. The full code is fairly simple:
 
 .. code-block:: python
 
-    class TagFunction(Protocol):
-        def __call__(self, *args: Decoded | Interpolation) -> Any:
-            ...
+    from types import Template, Interpolation
 
-Because of subclassing, the signature for ``mytag`` can of course be widened to
-the following, at the cost of losing some type specificity:
+    def convert(value: Any, conv: Literal["a", "r", "s"] | None) -> Any:
+        if conv == "a":
+            return ascii(value)
+        elif conv == "r":
+            return repr(value)
+        elif conv == "s":
+            return str(value)
+        return value
```

Comment:
@davepeck 
I don't have a strong opinion, but should this have been a `match case` instead of an `if`?
Or do you keep the `if` statement purposely?

Reply:
I felt the same at first, and also thought to change the `elif` to `if`. But I decided this is a simple, normal treatment. Structural pattern matching is still odd for some and would distract from the point made here.

Reply:
@koxudaxi @pauleveritt 

Just because I'm curious _and_ because I love bike sheds:

Does the python team have a preference here?

I'm also curious what everyone's personal preference would be of the following five options:

One:

```python
def convert1(value: Any, conv: Literal["a", "r", "s"] | None) -> Any:
    if conv == "a":
        return list(value)
    elif conv == "r":
        return tuple(value)
    elif conv == "s":
        return str(value)
    else:
        return value
```

Two:

```python
def convert2(value: Any, conv: Literal["a", "r", "s"] | None) -> Any:
    if conv == "a":
        return list(value)
    elif conv == "r":
        return tuple(value)
    elif conv == "s":
        return str(value)
    return value
```

Three:

```python
def convert3(value: Any, conv: Literal["a", "r", "s"] | None) -> Any:
    if conv == "a":
        return list(value)
    if conv == "r":
        return tuple(value)
    if conv == "s":
        return str(value)
    return value
```

Four:

```python
def convert4(value: Any, conv: Literal["a", "r", "s"] | None) -> Any:
    match conv:
        case "a":
            return list(value)
        case "r":
            return tuple(value)
        case "s":
            return str(value)
        case _:
            return value
```

Five:

```python
def convert5(value: Any, conv: Literal["a", "r", "s"] | None) -> Any:
    match conv:
        case "a":
            return list(value)
        case "r":
            return tuple(value)
        case "s":
            return str(value)
    return value
```

That's before considering:

1. Variants that have a single `return` statement (aka `return converted_value`)
2. Variants that don't allow `conv` to be `None` (and therefore can return `str` rather than `Any`)

I admit that I think too much about nonsense like this. :-)

Reply:
@davepeck 
I prefer `Two` style because I  would avoid too many indents.
But I'm trying to catch up on the new standard `Structural Pattern Matching`.
`Four` is the new standard for us?
I applied the style in my daily job code.
I feel it's too many indents 😅 I think a better way to use the style.
Here is my code. 
Look good.
https://github.com/Mirascope/mirascope/blob/265c5f7c688957653ea6b1d64f432ca003f7007a/mirascope/beta/openai/realtime/realtime.py#L233-L280

This has too many indents 🤔 
https://github.com/Mirascope/mirascope/blob/265c5f7c688957653ea6b1d64f432ca003f7007a/mirascope/beta/openai/realtime/realtime.py#L183-L215

Reply:
I prefer (3).

Reply:
I'm team (3) as well.